### PR TITLE
Proposal: Hierarchical Queue on Capacity Plugin

### PR DESCRIPTION
**Pull request by Rui-Gan**
issue #3590 

Add MVP design for hierarchical Queue on Capacity Plugin

### CODE REVIEW

**Comment by googs1025** [comment]
Hi, we can solve the DCO problem first. You can use `git commit -s xxxx`

File: docs/design/hierarchical-queue-on-capacity-plugin.md [hunk 1]
```diff
@@ -0,0 +1,118 @@
+# Hierarchical Queue on Capacity Plugin
+
+## Motivation
+
+In the context of cloud-native AI task scheduling, fair scheduling and resource utilization are two critical concerns for users. The Volcano community has already developed the elastic queue management plugin, known as capacity, to support fine-grained resource borrowing/lending and queue management, thereby improving resource utilization. However, in real-world scenarios, queues are often hierarchical, corresponding to the hierarchical organizational structure of company teams. To better align with practical queue usage scenarios and further enhance resource utilization for AI task scheduling, Volcano needs to build upon the existing capacity plugin and introduce hierarchical queue management capabilities.
+
+## In Scope
+
+- Enhancing the elastic queue management capabilities of the capacity plugin to support hierarchical queues.
```
Comment (googs1025) [inline docs/design/hierarchical-queue-on-capacity-plugin.md:9]: Will this proposal add a new CR object for users to create? Maybe ChildQueue? Or is this completely internal logic? Will it not be exposed to users?

File: docs/design/hierarchical-queue-on-capacity-plugin.md [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+<div align="center"><img width="582" height="393" src="images/hierarchical-queue-on-capacity-plugin.png" /></div>
+
+When job A performs reclaim or preempt actions, the system should first consider job B (belonging to the same parent queue a), and then consider job C (belonging to the same parent queue root).
+
+## Design detail
+
+### Data structure modifications
+
+- Extend the queueAttr struct in capacity plugin to include two new fields:
```
Comment (googs1025) [inline docs/design/hierarchical-queue-on-capacity-plugin.md:43]: Do we have a limit on the number of sub- child?

File: docs/design/hierarchical-queue-on-capacity-plugin.md (hunk 1 above)
Comment (Rui-Gan) [reply docs/design/hierarchical-queue-on-capacity-plugin.md:9]: This proposal only involves internal logic changes within the Capacity plugin. It does not introduce any new CR. 

The existing Queue CR already has a parent field. Internally, the Capacity plugin will establish the hierarchical relationships between queues based on the parent field of all the Queues.

File: docs/design/hierarchical-queue-on-capacity-plugin.md (hunk 2 above)
Comment (Rui-Gan) [reply docs/design/hierarchical-queue-on-capacity-plugin.md:43]: Currently, there is no limit. If you have any suggestions or recommendations regarding limiting the number of sub-child, I would greatly appreciate your input.

File: docs/design/hierarchical-queue-on-capacity-plugin.md (hunk 1 above)
Comment (googs1025) [reply docs/design/hierarchical-queue-on-capacity-plugin.md:9]: Will there be any changes to the Queue CR API? If so, we should also show it in the documentation.

Comment (Rui-Gan) [reply docs/design/hierarchical-queue-on-capacity-plugin.md:9]: It does not require any changes to the Queue CR API. From a user's perspective, the way they create and manage Queue CRs will remain the same.

**Comment by googs1025** [comment]
let's squash  to one commit

**Comment by Rui-Gan** [comment]
> let's squash to one commit

fix it

File: docs/design/hierarchical-queue-on-capacity-plugin.md [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+  - For example, in the case of `AddJobEnqueueableFn`, it should directly return `util.Reject`, preventing the job from being enqueued. No scheduling should be performed in situations that do not align with the previously described conditions. Users need to adjust the queues based on the printed error information until normal scheduling can be resumed.
+- `AddQueueOrderFn`: Prioritize leaf queues (queues without child queues) when considering queue ordering. For two leaf queues, the ordering will be determined by considering their parent queues layer by layer, starting from the root queue.
+- `AddAllocatableFn`: Ensure that only leaf queues can schedule jobs/podgroups.
+- `BuildVictimsPriorityQueue`: The current implementation of this function relies on `TaskOrderFn` and `JobOrderFn` for sorting victims. Two new functions, `VictimTaskOrderFn` and `VictimJobOrderFn`, will be introduced specifically for sorting victims in the context of hierarchical queues. 
+  
+  ```go
+  func (ssn *Session) BuildVictimsPriorityQueue(victims []*api.TaskInfo, preemptor *api.TaskInfo) *util.PriorityQueue {
```
Comment (Monokaix) [inline docs/design/hierarchical-queue-on-capacity-plugin.md:98]: Seems that this logic should exist in `AddReclaimableFn`, right?

File: docs/design/hierarchical-queue-on-capacity-plugin.md [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+- When creating a queue, the following logic will be added to the ValidatingAdmissionWebhook:
+  - If the parent queue of the queue to be created is currently scheduling jobs/podgroups, the creation of the new queue will not be allowed.
+  - The capability of the parent queue of the queue to be created must be greater than or equal to the capability of the new queue, and the deserved/guarantee of the parent queue must be greater than or equal to the sum of all its child queues (including the queue to be created). If this condition is not satisfied, the creation of the new queue will not be allowed.
+- The above modifications are primarily applicable when `EnabledHierarchy` is set to true. If the capacity plugin does not require hierarchical queue management, the existing implementations of these functions will be retained.
+
+### Limitations:
+
+- The current design does not support scheduling jobs/podgroups to non-leaf queues. Only leaf queues can directly schedule and allocate resources to jobs/podgroups.
+- Design relevant vcctl commands, such as commands to obtain the child queues of a specific queue or commands to retrieve the entire hierarchical queue structure.
```
Comment (Monokaix) [inline docs/design/hierarchical-queue-on-capacity-plugin.md:131]: This seems not a limitation, it should be implemented, we can put this into `Key function modifications` part, and if this feature's priority is relatively low, we can implement it in stages.

File: docs/design/hierarchical-queue-on-capacity-plugin.md [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+    
+    queueOpts map[api.QueueID]*queueAttr
+    // Arguments given for the plugin
+    pluginArguments framework.Arguments
+  }
+  ```
+
+**Note:** When the scheduler has the capacity plugin enabled and `EnabledHierarchy` is set to true, a root queue with queueID `root` will be automatically created. In the `OpenSession` function, the `Deserved` and `Capability` fields of the root queue are updated each time based on `ssn.TotalResource`.
+
+### Key function modifications
```
Comment (Monokaix) [inline docs/design/hierarchical-queue-on-capacity-plugin.md:88]: Change Implement is better.

File: docs/design/hierarchical-queue-on-capacity-plugin.md [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+- `OnSessionOpen`: Check that the sum of the deserved/guarantee values of each queue's child queues does not exceed the deserved/guarantee of the parent queue, and that the capability of any child queue does not exceed that of the parent queue. 
+  - If the check reveals non-compliance with the aforementioned conditions, an error message should be printed, and the relevant functions for the session should return a disallowed status. 
+  - For example, in the case of `AddJobEnqueueableFn`, it should directly return `util.Reject`, preventing the job from being enqueued. No scheduling should be performed in situations that do not align with the previously described conditions. Users need to adjust the queues based on the printed error information until normal scheduling can be resumed.
+- `AddQueueOrderFn`: Prioritize leaf queues (queues without child queues) when considering queue ordering. For two leaf queues, the ordering will be determined by considering their parent queues layer by layer, starting from the root queue.
```
Comment (Monokaix) [inline docs/design/hierarchical-queue-on-capacity-plugin.md:93]: There are many limitation and validation in hierarchical queue, some of the validation mentioned in the document are in plugins, such as AddQueueOrderFn here, and some are in webhooks, such as limitations part below. It is best to perform validation in a unified place. You need to clarify where to put it.

File: docs/design/hierarchical-queue-on-capacity-plugin.md [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+
+In the context of cloud-native AI task scheduling, fair scheduling and resource utilization are two critical concerns for users. The Volcano community has already developed the elastic queue management plugin, known as capacity, to support fine-grained resource borrowing/lending and queue management, thereby improving resource utilization. However, in real-world scenarios, queues are often hierarchical, corresponding to the hierarchical organizational structure of company teams. To better align with practical queue usage scenarios and further enhance resource utilization for AI task scheduling, Volcano needs to build upon the existing capacity plugin and introduce hierarchical queue management capabilities.
+
+## In Scope
+
+- Enhancing the elastic queue management capabilities of the capacity plugin to support hierarchical queues.
+
+## Out of Scope
+
+- Extending the hierarchical queues of any other plugins.
```
Comment (hwdef) [inline docs/design/hierarchical-queue-on-capacity-plugin.md:13]: If the capacity plugin is not enabled, will the hierarchical queue still be valid? For example, is it possible to submit tasks to a subqueue?

File: docs/design/hierarchical-queue-on-capacity-plugin.md [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+When job A performs reclaim or preempt actions, the system should first consider job B (belonging to the same parent queue a), and then consider job C (belonging to the same parent queue root).
+
+## Design detail
+
+### Data structure modifications
+
+- Extend the queueAttr struct in capacity plugin to include two new fields:
+  - `parents`: Store all the parent queues of the current queue, including the root queue.
+  - `children`: Store all the child queues of the current queue.
```
Comment (hwdef) [inline docs/design/hierarchical-queue-on-capacity-plugin.md:45]: Why store all parent and child. Is it possible to store only one layer, that is, the parent queue and the child queue?

File: docs/design/hierarchical-queue-on-capacity-plugin.md (hunk 7 above)
Comment (Rui-Gan) [reply docs/design/hierarchical-queue-on-capacity-plugin.md:13]: If the Capacity plugin is not enabled, the hierarchical queue feature will not take effect; however, it's important to note that the DRF plugin also supports hierarchical queues, and users can utilize the hierarchical queue functionality provided by other plugins.
When a task is submitted to a child queue, the Capacity plugin does not perform any validation; it primarily depends on the validation of other plugins.

File: docs/design/hierarchical-queue-on-capacity-plugin.md (hunk 8 above)
Comment (Rui-Gan) [reply docs/design/hierarchical-queue-on-capacity-plugin.md:72]: When the resources of a queue change, it is possible to traverse all parent queues to perform updates; when conducting quota validation on a queue, the corresponding validation can be carried out through all child queues. Storing all parent and child queues is mainly used to establish the hierarchical relationship between queues.
It's important to note that child queues are only one level deep. For example, in the diagram from story 4, the child queues of the root queue are child queue a and child queue b, and do not include subchild queue a.

File: docs/design/hierarchical-queue-on-capacity-plugin.md (hunk 3 above)
Comment (Rui-Gan) [reply docs/design/hierarchical-queue-on-capacity-plugin.md:108]: AddReclaimableFn is used to select victims, while VictimTaskOrderFn and VictimJobOrderFn are used to sort the victims.

File: docs/design/hierarchical-queue-on-capacity-plugin.md (hunk 7 above)
Comment (hwdef) [reply docs/design/hierarchical-queue-on-capacity-plugin.md:13]: Can this vcjob be submitted normally?

Comment (Rui-Gan) [reply docs/design/hierarchical-queue-on-capacity-plugin.md:13]: yes

**Pull request merged**